File: backend/update-service/config/tableAccess.js
// updateservice/config/tableAccess.js
import bcryptjs from "bcryptjs";
import { fetchRecord } from "../utils/readService.js";

export default {
  // === STUDENT ===
  student: {
    /**
     * Hook executed before updating a student record.
     * Handles permission checks and password updates.
     *
     * @param {string} token - Authorization token of the user making the request.
     * @param {Object} user - The currently authenticated user.
     * @param {number|string} id - ID of the student to update.
     * @param {Object} data - Data to update, may include newPassword and currentPassword.
     * @returns {Promise<Object>} Object containing `allowed` (boolean) and `modifiedData` or `error`.
     */
    beforeUpdate: async (token, user, id, data) => {
      const isAdmin = user.type === "admin";
      const isSelf = user.type === "student" && user.id === parseInt(id);

      if (!token) return { allowed: false, error: "Missing authorization token" };
      if (!isAdmin && !isSelf) return { allowed: false, error: "Forbidden" };

      // === Password change handling ===
      if (data.newPassword) {
        let currentHash = null;

        // Admin can update without current password
        if (isAdmin) {
          const rows = await fetchRecord("student", id, false, user, token);
          if (!rows || rows.length === 0)
            return { allowed: false, error: "Student not found" };
        }

        // Self-update requires current password verification
        if (!isAdmin && isSelf) {
          if (!data.currentPassword)
            return { allowed: false, error: "Current password required" };

          const rows = await fetchRecord("student", id, true, user, token);
          if (!rows || rows.length === 0)
            return { allowed: false, error: "Student not found" };

          currentHash = rows[0].password;
          const match = await bcryptjs.compare(data.currentPassword, currentHash);
          if (!match)
            return { allowed: false, error: "Current password is incorrect" };
        }

        // Hash new password and replace in data
        data.password = await bcryptjs.hash(data.newPassword, 10);
        delete data.currentPassword;
        delete data.newPassword;
      }

      return { allowed: true, modifiedData: data };
    },
  },

  // === TEACHER ===
  teacher: {
    /**
     * Hook executed before updating a teacher record.
     * Handles permission checks and password updates.
     *
     * @param {string} token - Authorization token of the user making the request.
     * @param {Object} user - The currently authenticated user.
     * @param {number|string} id - ID of the teacher to update.
     * @param {Object} data - Data to update, may include newPassword and currentPassword.
     * @returns {Promise<Object>} Object containing `allowed` (boolean) and `modifiedData` or `error`.
     */
    beforeUpdate: async (token, user, id, data) => {
      const isAdmin = user.type === "admin";
      const isSelf = user.type === "teacher" && user.id === parseInt(id);

      if (!token) return { allowed: false, error: "Missing authorization token" };
      if (!isAdmin && !isSelf) return { allowed: false, error: "Forbidden" };

      if (data.newPassword) {
        let currentHash = null;

        if (isAdmin) {
          const rows = await fetchRecord("teacher", id, false, user, token);
          if (!rows || rows.length === 0)
            return { allowed: false, error: "Teacher not found" };
        }

        if (!isAdmin && isSelf) {
          if (!data.currentPassword)
            return { allowed: false, error: "Current password required" };

          const rows = await fetchRecord("teacher", id, true, user, token);
          if (!rows || rows.length === 0)
            return { allowed: false, error: "Teacher not found" };

          currentHash = rows[0].password;
          const match = await bcryptjs.compare(data.currentPassword, currentHash);
          if (!match)
            return { allowed: false, error: "Current password is incorrect" };
        }

        data.password = await bcryptjs.hash(data.newPassword, 10);
        delete data.currentPassword;
        delete data.newPassword;
      }

      return { allowed: true, modifiedData: data };
    },
  },

  // === ADMIN ===
  admin: {
    /**
     * Hook executed before updating an admin record.
     * Admins can only update their own account and must verify current password for changes.
     *
     * @param {string} token - Authorization token of the user making the request.
     * @param {Object} user - The currently authenticated user.
     * @param {number|string} id - ID of the admin to update.
     * @param {Object} data - Data to update, may include newPassword and currentPassword.
     * @returns {Promise<Object>} Object containing `allowed` (boolean) and `modifiedData` or `error`.
     */
    beforeUpdate: async (token, user, id, data) => {
      const isSelf = user.id === parseInt(id);
      if (!isSelf)
        return { allowed: false, error: "Admins can only update their own account" };

      if (data.newPassword) {
        if (!data.currentPassword)
          return { allowed: false, error: "Current password required" };

        const rows = await fetchRecord("admin", id, true, user, token);
        if (!rows || rows.length === 0)
          return { allowed: false, error: "Admin not found" };

        const match = await bcryptjs.compare(data.currentPassword, rows[0].password);
        if (!match)
          return { allowed: false, error: "Current password incorrect" };

        data.password = await bcryptjs.hash(data.newPassword, 10);
        delete data.currentPassword;
        delete data.newPassword;
      }

      return { allowed: true, modifiedData: data };
    },
  },

  // === COURSE ===
  course: {
    /**
     * Hook executed before updating a course record.
     * Only admins can update courses, and only specific fields are allowed.
     *
     * @param {string} token - Authorization token of the user making the request.
     * @param {Object} user - The currently authenticated user.
     * @param {number|string} id - ID of the course to update.
     * @param {Object} data - Data to update (name, teacherId, credit_hours).
     * @returns {Promise<Object>} Object containing `allowed` (boolean) and `modifiedData` or `error`.
     */
    beforeUpdate: async (token, user, id, data) => {
      if (user.type !== "admin")
        return { allowed: false, error: "Only admins can update courses" };

      const rows = await fetchRecord("course", id, false, user, token);
      if (!rows || rows.length === 0)
        return { allowed: false, error: "Course not found" };

      const allowed = ["name", "teacherId", "credit_hours"];
      const filtered = Object.fromEntries(
        Object.entries(data).filter(([k]) => allowed.includes(k))
      );

      if (Object.keys(filtered).length === 0)
        return { allowed: false, error: "No valid fields to update" };

      return { allowed: true, modifiedData: filtered };
    },
  },

  // === ENROLLMENT ===
  enrollment: {
    /**
     * Hook executed before updating an enrollment record.
     * Enrollment records are immutable and cannot be updated.
     *
     * @returns {Object} Object with `allowed` set to false and an `error` message.
     */
    beforeUpdate: async () => ({
      allowed: false,
      error: "Enrollments cannot be updated",
    }),
  },
};

File: djangoBackend/update-service/updateserv/config/table_access.py
# update config
import httpx
import json
import bcrypt
from updateserv.utils.read_service import fetch_record
from updateserv.utils.uniqueness import check_global_uniqueness_on_update
from updateserv.exceptions import UniquenessError


READ_SERVICE_URL = "http://localhost:4000/read/api/get-data" 

class TableAccess:
    async def student_before_update(self, token, user, record_id, data):
        is_admin = user["type"] == "admin"
        is_self = user["type"] == "student" and user["id"] == int(record_id)

        if not token:
            return False, "Missing authorization token", None
        if not is_admin and not is_self:
            return False, "Forbidden", None
        
        await check_global_uniqueness_on_update(
        user=user,
        token=token,
        data=data,
        record_id=record_id,
        current_table="student",
        )

        # password logic
        if "newPassword" in data:
            # admin updates without current password
            if is_admin:
                rows = await fetch_record("student", record_id, False, user, token)
                if not rows:
                    return False, "Student not found", None

            # self update = verify current password
            if is_self:
                if "currentPassword" not in data:
                    return False, "Current password required", None

                rows = await fetch_record("student", record_id, True, user, token)
                if not rows:
                    return False, "Student not found", None

                current_hash = rows[0]["password"]
                if not bcrypt.checkpw(data["currentPassword"].encode(), current_hash.encode()):
                    return False, "Current password is incorrect", None

            # hash new password
            hashed = bcrypt.hashpw(data["newPassword"].encode('utf-8'), bcrypt.gensalt(rounds=10)).decode('utf-8')
            data["password"] = hashed
            data.pop("newPassword", None)
            data.pop("currentPassword", None)

        return True, None, data

    async def teacher_before_update(self, token, user, record_id, data):
        is_admin = user["type"] == "admin"
        is_self = user["type"] == "teacher" and user["id"] == int(record_id)

        if not token:
            return False, "Missing authorization token", None
        if not is_admin and not is_self:
            return False, "Forbidden", None
        
        await check_global_uniqueness_on_update(
        user=user,
        token=token,
        data=data,
        record_id=record_id,
        current_table="teacher",
        )

        if "newPassword" in data:
            if is_admin:
                rows = await fetch_record("teacher", record_id, False, user, token)
                if not rows:
                    return False, "Teacher not found", None

            if is_self:
                if "currentPassword" not in data:
                    return False, "Current password required", None

                rows = await fetch_record("teacher", record_id, True, user, token)
                if not rows:
                    return False, "Teacher not found", None

                current_hash = rows[0]["password"]
                if not bcrypt.checkpw(data["currentPassword"].encode(), current_hash.encode()):
                    return False, "Current password is incorrect", None

            hashed = bcrypt.hashpw(data["newPassword"].encode(), bcrypt.gensalt()).decode()
            data["password"] = hashed
            data.pop("newPassword", None)
            data.pop("currentPassword", None)

        return True, None, data


    async def admin_before_update(self, token, user, record_id, data):
        if not token:
            return False, "Missing authorization token", None

        if user["id"] != int(record_id):
            return False, "Admins can only update their own account", None

        #  FORCE uniqueness check (admin was skipping this)
        await check_global_uniqueness_on_update(
            user=user,
            token=token,
            data=data,
            record_id=record_id,
            current_table="admin",
        )

        #  Password logic (unchanged)
        if "newPassword" in data:
            if "currentPassword" not in data:
                return False, "Current password required", None

            rows = await fetch_record("admin", record_id, True, user, token)
            if not rows:
                return False, "Admin not found", None

            current_hash = rows[0]["password"]
            if not bcrypt.checkpw(
                data["currentPassword"].encode(),
                current_hash.encode()
            ):
                return False, "Current password incorrect", None

            hashed = bcrypt.hashpw(
                data["newPassword"].encode(),
                bcrypt.gensalt()
            ).decode()

            data["password"] = hashed
            data.pop("newPassword", None)
            data.pop("currentPassword", None)

        return True, None, data


    async def course_before_update(self, token, user, record_id, data):
        if user["type"] != "admin":
            return False, "Only admins can update courses", None

        rows = await fetch_record("course", record_id, False, user, token)
        if not rows:
            return False, "Course not found", None
        
        async with httpx.AsyncClient(timeout=10.0) as client:
            try:
                response = await client.get(
                    READ_SERVICE_URL,
                    params={"tableName": "course"},
                    headers={
                        "X-User": json.dumps(user),
                        "Authorization": token,
                    },
                )
                response.raise_for_status()
                existing = response.json().get('data', [])

                for obj in existing:
                    if obj.get("isDeleted") is not False:
                        continue

                    if obj.get("id") == int(record_id):
                        continue

                    if obj.get('name') == data.get('name') and obj.get('isDeleted') is False:
                        raise UniquenessError(f"Another course with the name '{data.get('name')}' already exists")

            except httpx.HTTPError:
                raise Exception("Failed to communicate with Read Service during course validation.")


        allowed = ["name", "teacherId", "credit_hours"]
        filtered = {k: v for k, v in data.items() if k in allowed}

        if not filtered:
            return False, "No valid fields to update", None

        return True, None, filtered

    async def enrollment_before_update(self, token, user, record_id, data):
        return False, "Enrollments cannot be updated", None


table_access = TableAccess()
